File: py/packages/protocol/src/clamator_protocol/envelope.py
from __future__ import annotations
import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Union

SERVICE_RE = re.compile(r"^[a-z][a-z0-9-]*$")
METHOD_RE = re.compile(r"^[a-z][a-zA-Z0-9-]*$")


class EnvelopeKind(Enum):
    REQUEST = "request"
    NOTIFICATION = "notification"
    SUCCESS_RESPONSE = "success"
    ERROR_RESPONSE = "error"


RpcId = Union[str, int]


@dataclass(frozen=True)
class RequestEnvelope:
    service: str
    method: str
    full_method: str
    params: Any
    id: RpcId
    raw: dict[str, Any] = field(repr=False)
    kind: EnvelopeKind = EnvelopeKind.REQUEST


@dataclass(frozen=True)
class NotificationEnvelope:
    service: str
    method: str
    full_method: str
    params: Any
    raw: dict[str, Any] = field(repr=False)
    kind: EnvelopeKind = EnvelopeKind.NOTIFICATION


@dataclass(frozen=True)
class SuccessResponseEnvelope:
    id: RpcId
    result: Any
    kind: EnvelopeKind = EnvelopeKind.SUCCESS_RESPONSE


@dataclass(frozen=True)
class ErrorResponseEnvelope:
    id: RpcId | None
    error: dict[str, Any]
    kind: EnvelopeKind = EnvelopeKind.ERROR_RESPONSE


Envelope = Union[
    RequestEnvelope, NotificationEnvelope, SuccessResponseEnvelope, ErrorResponseEnvelope
]


def _invalid(msg: str) -> ValueError:
    return ValueError(f"-32600 Invalid Request: {msg}")


def parse_envelope(value: Any) -> Envelope:
    if isinstance(value, list):
        raise _invalid("batch requests not supported")
    if not isinstance(value, dict):
        raise _invalid("not an object")
    if value.get("jsonrpc") != "2.0":
        raise _invalid('jsonrpc must equal "2.0"')

    has_method = isinstance(value.get("method"), str)
    has_result = "result" in value
    has_error = "error" in value
    has_id = "id" in value and value["id"] is not None

    if has_method:
        full_method = value["method"]
        if "." not in full_method:
            raise _invalid('method must be "<service>.<method>"')
        service, _, method = full_method.partition(".")
        if not service or not method:
            raise _invalid('method must be "<service>.<method>"')
        if not SERVICE_RE.match(service):
            raise _invalid("invalid service segment")
        if not METHOD_RE.match(method):
            raise _invalid("invalid method segment")
        if has_id:
            rpc_id = value["id"]
            if not isinstance(rpc_id, (str, int)):
                raise _invalid("id must be string or number")
            return RequestEnvelope(
                service=service, method=method, full_method=full_method,
                params=value.get("params", {}), id=rpc_id, raw=value,
            )
        return NotificationEnvelope(
            service=service, method=method, full_method=full_method,
            params=value.get("params", {}), raw=value,
        )

    if has_result and not has_error:
        if not has_id:
            raise _invalid("response must have id")
        return SuccessResponseEnvelope(id=value["id"], result=value["result"])

    if has_error and not has_result:
        err = value.get("error")
        if not isinstance(err, dict) or not isinstance(err.get("code"), int) or not isinstance(err.get("message"), str):
            raise _invalid("error must have numeric code + string message")
        rpc_id = value["id"] if has_id else None
        return ErrorResponseEnvelope(
            id=rpc_id,
            error={"code": err["code"], "message": err["message"], "data": err.get("data")},
        )

    raise _invalid("envelope must be request, notification, or response")


def build_request(full_method: str, params: Any, rpc_id: RpcId) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "method": full_method, "params": params, "id": rpc_id}


def build_notification(full_method: str, params: Any) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "method": full_method, "params": params}


def build_success_response(rpc_id: RpcId, result: Any) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": rpc_id, "result": result}


def build_error_response(rpc_id: RpcId | None, code: int, message: str, data: Any = None) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": rpc_id, "error": {"code": code, "message": message, "data": data}}

File: ts/packages/protocol/src/envelope.ts
export const SERVICE_RE = /^[a-z][a-z0-9-]*$/;
// Method names must be valid identifiers in both TS and Py output of the codegen.
// Hyphens were previously allowed in the wire-side regex but broke codegen Py
// emission ("def group-cancel" is invalid syntax). Tightened in v0.1.6 to keep
// the wire-side and codegen-side validity definitions in agreement.
export const METHOD_RE = /^[a-z][a-zA-Z0-9]*$/;

// Plain string enum (NOT `const enum` — incompatible with `isolatedModules: true`).
export enum EnvelopeKind {
	Request = 'request',
	Notification = 'notification',
	SuccessResponse = 'success',
	ErrorResponse = 'error',
}

export type RpcId = string | number;

export interface RequestEnvelope {
	kind: EnvelopeKind.Request;
	service: string;
	method: string;
	fullMethod: string;
	params: unknown;
	id: RpcId;
	raw: Record<string, unknown>;
}

export interface NotificationEnvelope {
	kind: EnvelopeKind.Notification;
	service: string;
	method: string;
	fullMethod: string;
	params: unknown;
	raw: Record<string, unknown>;
}

export interface SuccessResponseEnvelope {
	kind: EnvelopeKind.SuccessResponse;
	id: RpcId;
	result: unknown;
}

export interface ErrorResponseEnvelope {
	kind: EnvelopeKind.ErrorResponse;
	id: RpcId | null;
	error: { code: number; message: string; data: unknown };
}

export type Envelope =
	| RequestEnvelope
	| NotificationEnvelope
	| SuccessResponseEnvelope
	| ErrorResponseEnvelope;

class InvalidRequest extends Error {
	readonly code = -32600;
	constructor(msg: string) {
		super(`-32600 Invalid Request: ${msg}`);
	}
}

export function parseEnvelope(value: unknown): Envelope {
	if (Array.isArray(value)) throw new InvalidRequest('batch requests not supported');
	if (typeof value !== 'object' || value === null) throw new InvalidRequest('not an object');
	const obj = value as Record<string, unknown>;
	if (obj.jsonrpc !== '2.0') throw new InvalidRequest('jsonrpc must equal "2.0"');

	const hasMethod = typeof obj.method === 'string';
	const hasResult = 'result' in obj;
	const hasError = 'error' in obj;
	const hasId = 'id' in obj && obj.id !== null && obj.id !== undefined;

	if (hasMethod) {
		const fullMethod = obj.method as string;
		const dot = fullMethod.indexOf('.');
		if (dot <= 0 || dot === fullMethod.length - 1)
			throw new InvalidRequest('method must be "<service>.<method>"');
		const service = fullMethod.slice(0, dot);
		const method = fullMethod.slice(dot + 1);
		if (!SERVICE_RE.test(service)) throw new InvalidRequest('invalid service segment');
		if (!METHOD_RE.test(method)) throw new InvalidRequest('invalid method segment');
		if (hasId) {
			const id = obj.id as RpcId;
			if (typeof id !== 'string' && typeof id !== 'number')
				throw new InvalidRequest('id must be string or number');
			return {
				kind: EnvelopeKind.Request,
				service, method, fullMethod,
				params: obj.params ?? {},
				id, raw: obj,
			};
		}
		return {
			kind: EnvelopeKind.Notification,
			service, method, fullMethod,
			params: obj.params ?? {},
			raw: obj,
		};
	}

	if (hasResult && !hasError) {
		if (!hasId) throw new InvalidRequest('response must have id');
		return {
			kind: EnvelopeKind.SuccessResponse,
			id: obj.id as RpcId,
			result: obj.result,
		};
	}

	if (hasError && !hasResult) {
		const err = obj.error as Record<string, unknown> | undefined;
		if (!err || typeof err.code !== 'number' || typeof err.message !== 'string')
			throw new InvalidRequest('error must have numeric code + string message');
		return {
			kind: EnvelopeKind.ErrorResponse,
			id: (hasId ? (obj.id as RpcId) : null),
			error: { code: err.code, message: err.message, data: err.data ?? null },
		};
	}

	throw new InvalidRequest('envelope must be request, notification, or response');
}

export function buildSuccessResponse(id: RpcId, result: unknown): Record<string, unknown> {
	return { jsonrpc: '2.0', id, result };
}

export function buildErrorResponse(
	id: RpcId | null,
	code: number, message: string, data: unknown = null,
): Record<string, unknown> {
	return { jsonrpc: '2.0', id, error: { code, message, data } };
}

export function buildRequest(fullMethod: string, params: unknown, id: RpcId): Record<string, unknown> {
	return { jsonrpc: '2.0', method: fullMethod, params, id };
}

export function buildNotification(fullMethod: string, params: unknown): Record<string, unknown> {
	return { jsonrpc: '2.0', method: fullMethod, params };
}
